src/cloud-providers/gcp/provider.go

```diff
@@ -0,0 +1,191 @@
+		if err != nil {
+			return nil, fmt.Errorf("configuration error when using creds: %s", err)
+		}
+		provider.instancesClient, err = compute.NewInstancesRESTClient(context.TODO(), option.WithCredentials(creds))
+		if err != nil {
+			return nil, fmt.Errorf("NewInstancesRESTClient with credentials error: %s", err)
+		}
+	} else {
+		var err error
+		provider.instancesClient, err = compute.NewInstancesRESTClient(context.TODO())
+		if err != nil {
+			return nil, fmt.Errorf("NewInstancesRESTClient error: %s", err)
+		}
+	}
+	return provider, nil
+}
+
+func getIPs(instance *computepb.Instance) ([]netip.Addr, error) {
+	var podNodeIPs []netip.Addr
+	for _, nic := range instance.GetNetworkInterfaces() {
+		for _, access := range nic.GetAccessConfigs() {
+			ipStr := access.GetNatIP()
+			ip, err := netip.ParseAddr(ipStr)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse pod node IP %q: %w", ipStr, err)
+			}
+			podNodeIPs = append(podNodeIPs, ip)
+			logger.Printf("%s", ip.String())
+		}
+	}
+	return podNodeIPs, nil
+}
+
+func (p *gcpProvider) CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator, spec provider.InstanceTypeSpec) (*provider.Instance, error) {
+
+	instanceName := util.GenerateInstanceName(podName, sandboxID, maxInstanceNameLen)
+	logger.Printf("CreateInstance: name: %q", instanceName)
+
+	userData, err := cloudConfig.Generate()
+	if err != nil {
+		return nil, err
+	}
+
+	//Convert userData to base64
+	userDataEnc := base64.StdEncoding.EncodeToString([]byte(userData))
+	logger.Printf("userDataEnc:  %s", userDataEnc)
+
+	insertReq := &computepb.InsertInstanceRequest{
+		Project: p.serviceConfig.ProjectId,
+		Zone:    p.serviceConfig.Zone,
+		InstanceResource: &computepb.Instance{
+			Name: proto.String(instanceName),
+			Disks: []*computepb.AttachedDisk{
+				{
+					InitializeParams: &computepb.AttachedDiskInitializeParams{
+						DiskSizeGb:  proto.Int64(20),
+						SourceImage: proto.String(fmt.Sprintf("projects/%s/global/images/%s", p.serviceConfig.ProjectId, p.serviceConfig.ImageName)),
+						DiskType:    proto.String(fmt.Sprintf("zones/%s/diskTypes/pd-standard", p.serviceConfig.Zone)),
+					},
+					AutoDelete: proto.Bool(true),
+					Boot:       proto.Bool(true),
+					Type:       proto.String(computepb.AttachedDisk_PERSISTENT.String()),
+				},
+			},
+			Metadata: &computepb.Metadata{
+				Items: []*computepb.Items{
+					{
+						Key:   proto.String("user-data"),
+						Value: proto.String(userDataEnc),
+					},
+					{
+						Key:   proto.String("user-data-encoding"),
+						Value: proto.String("base64"),
+					},
+				},
+			},
+			MachineType: proto.String(fmt.Sprintf("zones/%s/machineTypes/%s", p.serviceConfig.Zone, p.serviceConfig.MachineType)),
+			NetworkInterfaces: []*computepb.NetworkInterface{
+				{
+					AccessConfigs: []*computepb.AccessConfig{
+						{
+							Name:        proto.String("External NAT"),
```

Comment:
IIUC this will create a VM with public IP address. The AWS implementation creates a VM with private IP by default but it allow users to choose to use public IP (see commit 39d69d494b982f008e0b1767469cf914b1fa835f). Other providers I don't know how it was implemented.

Reply:
@wainersm, Will the PodVM be able to communicate with external clusters within a private network? This is a genuine question.

Given that GKE is not supported at this point, I am concerned it may not add value currently.

Reply:
Hi @beraldoleal ! Not expert speaking, but I guess CAA running on an external cluster won't be able to talk to the VM within a private network unless you provide the correct routing.

I'm not saying that you should implement the private IP on this first version. I just wanted to caught your attention to the fact that some providers like AWS provides that flexibility, and you might be interested on doing the same on the follow on PRs.